### PR TITLE
[WIP][Scala] Module API Support

### PR DESCRIPTION
**Pull request by javelinjs**
Here's what I got for `examples/scripts/module/mnist_mlp.sh`:

```
[01:10:54] src/io/iter_mnist.cc:91: MNISTIter: load 60000 images, shuffle=1, shape=(10,784)
[01:10:54] src/io/iter_mnist.cc:91: MNISTIter: load 10000 images, shuffle=1, shape=(10,784)
2017-02-11 01:11:03,389 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - epoch 0 accuracy=0.81295
2017-02-11 01:11:11,801 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - epoch 1 accuracy=0.96045
2017-02-11 01:11:19,246 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - epoch 2 accuracy=0.97318333
2017-02-11 01:11:26,635 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - epoch 3 accuracy=0.9789
2017-02-11 01:11:34,581 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - epoch 4 accuracy=0.9823
2017-02-11 01:11:42,359 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - epoch 5 accuracy=0.98415
2017-02-11 01:11:51,417 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - epoch 6 accuracy=0.9867333
2017-02-11 01:11:59,458 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - epoch 7 accuracy=0.9877333
2017-02-11 01:12:07,118 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - epoch 8 accuracy=0.9887667
2017-02-11 01:12:16,234 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - epoch 9 accuracy=0.9888833
2017-02-11 01:12:23,087 [main] [ml.dmlc.mxnet.module.BaseModule] [INFO] - Epoch[0] Train-accuracy=0.13861667
2017-02-11 01:12:23,087 [main] [ml.dmlc.mxnet.module.BaseModule] [INFO] - Epoch[0] Time cost=6826
2017-02-11 01:12:23,365 [main] [ml.dmlc.mxnet.module.BaseModule] [INFO] - Epoch[0] Validation-accuracy=0.3904
2017-02-11 01:12:29,399 [main] [ml.dmlc.mxnet.module.BaseModule] [INFO] - Epoch[1] Train-accuracy=0.74293333
2017-02-11 01:12:29,400 [main] [ml.dmlc.mxnet.module.BaseModule] [INFO] - Epoch[1] Time cost=6035
2017-02-11 01:12:29,716 [main] [ml.dmlc.mxnet.module.BaseModule] [INFO] - Epoch[1] Validation-accuracy=0.8561
2017-02-11 01:12:37,259 [main] [ml.dmlc.mxnet.module.BaseModule] [INFO] - Epoch[2] Train-accuracy=0.89135
2017-02-11 01:12:37,259 [main] [ml.dmlc.mxnet.module.BaseModule] [INFO] - Epoch[2] Time cost=7543
2017-02-11 01:12:37,634 [main] [ml.dmlc.mxnet.module.BaseModule] [INFO] - Epoch[2] Validation-accuracy=0.9165
2017-02-11 01:12:45,638 [main] [ml.dmlc.mxnet.module.BaseModule] [INFO] - Epoch[3] Train-accuracy=0.92903334
2017-02-11 01:12:45,638 [main] [ml.dmlc.mxnet.module.BaseModule] [INFO] - Epoch[3] Time cost=8004
2017-02-11 01:12:45,929 [main] [ml.dmlc.mxnet.module.BaseModule] [INFO] - Epoch[3] Validation-accuracy=0.9394
2017-02-11 01:12:53,964 [main] [ml.dmlc.mxnet.module.BaseModule] [INFO] - Epoch[4] Train-accuracy=0.9471167
2017-02-11 01:12:53,965 [main] [ml.dmlc.mxnet.module.BaseModule] [INFO] - Epoch[4] Time cost=8036
2017-02-11 01:12:54,508 [main] [ml.dmlc.mxnet.module.BaseModule] [INFO] - Epoch[4] Validation-accuracy=0.9514
2017-02-11 01:13:01,534 [main] [ml.dmlc.mxnet.module.BaseModule] [INFO] - Epoch[5] Train-accuracy=0.95783335
2017-02-11 01:13:01,535 [main] [ml.dmlc.mxnet.module.BaseModule] [INFO] - Epoch[5] Time cost=7026
2017-02-11 01:13:02,036 [main] [ml.dmlc.mxnet.module.BaseModule] [INFO] - Epoch[5] Validation-accuracy=0.9585
2017-02-11 01:13:08,396 [main] [ml.dmlc.mxnet.module.BaseModule] [INFO] - Epoch[6] Train-accuracy=0.9645333
2017-02-11 01:13:08,396 [main] [ml.dmlc.mxnet.module.BaseModule] [INFO] - Epoch[6] Time cost=6359
2017-02-11 01:13:08,663 [main] [ml.dmlc.mxnet.module.BaseModule] [INFO] - Epoch[6] Validation-accuracy=0.9637
2017-02-11 01:13:14,277 [main] [ml.dmlc.mxnet.module.BaseModule] [INFO] - Epoch[7] Train-accuracy=0.96955
2017-02-11 01:13:14,277 [main] [ml.dmlc.mxnet.module.BaseModule] [INFO] - Epoch[7] Time cost=5614
2017-02-11 01:13:14,523 [main] [ml.dmlc.mxnet.module.BaseModule] [INFO] - Epoch[7] Validation-accuracy=0.9666
2017-02-11 01:13:20,128 [main] [ml.dmlc.mxnet.module.BaseModule] [INFO] - Epoch[8] Train-accuracy=0.9734667
2017-02-11 01:13:20,128 [main] [ml.dmlc.mxnet.module.BaseModule] [INFO] - Epoch[8] Time cost=5605
2017-02-11 01:13:20,376 [main] [ml.dmlc.mxnet.module.BaseModule] [INFO] - Epoch[8] Validation-accuracy=0.969
2017-02-11 01:13:26,104 [main] [ml.dmlc.mxnet.module.BaseModule] [INFO] - Epoch[9] Train-accuracy=0.9767
2017-02-11 01:13:26,104 [main] [ml.dmlc.mxnet.module.BaseModule] [INFO] - Epoch[9] Time cost=5728
2017-02-11 01:13:26,350 [main] [ml.dmlc.mxnet.module.BaseModule] [INFO] - Epoch[9] Validation-accuracy=0.9711
2017-02-11 01:13:26,353 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - Batch 0 acc: 1.0
2017-02-11 01:13:26,362 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - Batch 20 acc: 1.0
2017-02-11 01:13:26,370 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - Batch 40 acc: 0.9
2017-02-11 01:13:26,377 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - Batch 60 acc: 1.0
2017-02-11 01:13:26,383 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - Batch 80 acc: 0.9
2017-02-11 01:13:26,390 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - Batch 100 acc: 1.0
2017-02-11 01:13:26,396 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - Batch 120 acc: 0.9
2017-02-11 01:13:26,403 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - Batch 140 acc: 1.0
2017-02-11 01:13:26,409 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - Batch 160 acc: 1.0
2017-02-11 01:13:26,415 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - Batch 180 acc: 1.0
2017-02-11 01:13:26,423 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - Batch 200 acc: 1.0
2017-02-11 01:13:26,431 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - Batch 220 acc: 1.0
2017-02-11 01:13:26,437 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - Batch 240 acc: 1.0
2017-02-11 01:13:26,443 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - Batch 260 acc: 1.0
2017-02-11 01:13:26,449 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - Batch 280 acc: 1.0
2017-02-11 01:13:26,455 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - Batch 300 acc: 0.9
2017-02-11 01:13:26,461 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - Batch 320 acc: 1.0
2017-02-11 01:13:26,467 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - Batch 340 acc: 1.0
2017-02-11 01:13:26,473 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - Batch 360 acc: 1.0
2017-02-11 01:13:26,479 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - Batch 380 acc: 1.0
2017-02-11 01:13:26,485 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - Batch 400 acc: 1.0
2017-02-11 01:13:26,491 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - Batch 420 acc: 1.0
2017-02-11 01:13:26,497 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - Batch 440 acc: 1.0
2017-02-11 01:13:26,503 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - Batch 460 acc: 1.0
2017-02-11 01:13:26,509 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - Batch 480 acc: 1.0
2017-02-11 01:13:26,515 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - Batch 500 acc: 0.9
2017-02-11 01:13:26,521 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - Batch 520 acc: 1.0
2017-02-11 01:13:26,528 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - Batch 540 acc: 1.0
2017-02-11 01:13:26,536 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - Batch 560 acc: 1.0
2017-02-11 01:13:26,542 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - Batch 580 acc: 1.0
2017-02-11 01:13:26,548 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - Batch 600 acc: 1.0
2017-02-11 01:13:26,554 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - Batch 620 acc: 1.0
2017-02-11 01:13:26,559 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - Batch 640 acc: 1.0
2017-02-11 01:13:26,565 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - Batch 660 acc: 1.0
2017-02-11 01:13:26,570 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - Batch 680 acc: 1.0
2017-02-11 01:13:26,576 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - Batch 700 acc: 1.0
2017-02-11 01:13:26,582 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - Batch 720 acc: 1.0
2017-02-11 01:13:26,588 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - Batch 740 acc: 1.0
2017-02-11 01:13:26,594 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - Batch 760 acc: 1.0
2017-02-11 01:13:26,600 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - Batch 780 acc: 1.0
2017-02-11 01:13:26,606 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - Batch 800 acc: 0.8
2017-02-11 01:13:26,612 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - Batch 820 acc: 1.0
2017-02-11 01:13:26,617 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - Batch 840 acc: 1.0
2017-02-11 01:13:26,622 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - Batch 860 acc: 1.0
2017-02-11 01:13:26,628 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - Batch 880 acc: 0.9
2017-02-11 01:13:26,633 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - Batch 900 acc: 1.0
2017-02-11 01:13:26,640 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - Batch 920 acc: 1.0
2017-02-11 01:13:26,646 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - Batch 940 acc: 1.0
2017-02-11 01:13:26,652 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - Batch 960 acc: 0.9
2017-02-11 01:13:26,657 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - Batch 980 acc: 0.9
2017-02-11 01:13:27,236 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - Validation Accuracy: 0.9711
2017-02-11 01:13:27,490 [main] [ml.dmlc.mxnet.examples.module.MnistMlp] [INFO] - Scored accuracy = 0.9711
```

### CODE REVIEW

**Comment by javelinjs** [comment]
@CodingCat @Ldpe2G Please help to review.

File: scala-package/core/src/main/scala/ml/dmlc/mxnet/NDArray.scala [hunk 1]
```diff
@@ -305,29 +305,54 @@ object NDArray {
...
+      var idx = 0
+      val begin = Array.fill(retShape.length)(0)
+      val end = retShape.toArray
+      for (arr <- arrays) {
+        if (axis == 0) {
+          ret.slice(idx, idx + arr.shape(0)).set(arr)
+        } else {
+          begin(axis) = idx
+          end(axis) = idx + arr.shape(axis)
+          NDArray._crop_assign(Map("out" -> ret,
```
Comment (Ldpe2G) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/NDArray.scala]: Why not make these functions which start with "_" public ?

File: scala-package/examples/src/main/scala/ml/dmlc/mxnet/examples/module/MnistMlp.scala [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+    logger.info(s"Validation Accuracy: {}", accSum / accCnt.toFloat)
+
+    // evaluate on validation set with a evaluation metric
+    val (name, value) = mod.score(test, new Accuracy).get
+    logger.info("Scored {} = {}", name, value)
+  }
+
+  def main(args: Array[String]): Unit = {
+    val inst = new MnistMlp
+    val parser: CmdLineParser = new CmdLineParser(inst)
```
Comment (CodingCat) [inline scala-package/examples/src/main/scala/ml/dmlc/mxnet/examples/module/MnistMlp.scala]: is it a necessary claim of the variable type? (I mean we may only need to explicitly claim the local variable's type when it is hard to be detected at the first sight)

File: scala-package/examples/src/main/scala/ml/dmlc/mxnet/examples/module/MnistMlp.scala [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+    // prediction iterator API
+    var iBatch = 0
+    test.reset()
+    while (test.hasNext) {
+      val batch = test.next()
+      val preds = mod.predict(batch)
+      val predLabel: Array[Int] = NDArray.argmax_channel(preds(0)).toArray.map(_.toInt)
+      val label = batch.label(0).toArray.map(_.toInt)
+      val acc = (predLabel zip label).map { case (py, y) =>
```
Comment (CodingCat) [inline scala-package/examples/src/main/scala/ml/dmlc/mxnet/examples/module/MnistMlp.scala]: do we have some contract on the consistent usage of operator style or method style?

File: scala-package/core/src/main/scala/ml/dmlc/mxnet/module/DataParallelExecutorGroup.scala [hunk 4]
```diff
@@ -0,0 +1,695 @@
...
+ *                    In this case, many memory will be shared.
+ * @param inputTypes Default is `None`. When not `None`,
+ *                   can be used to specify the data type for each of the data/label inputs.
+ * @param fixedParamNames Indicate parameters to be fixed during training.
+ *                        Parameters in this list will not allocate space for gradient,
+ *                        nor do gradient calculation.
+ * @param gradReq Requirement for gradient accumulation. Can be 'write', 'add', or 'null',
+ *                be specified for each argument.
+ */
+class DataParallelExecutorGroup private[mxnet](
```
Comment (CodingCat) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/module/DataParallelExecutorGroup.scala]: maybe private to the package? it has a package-private constructor and I didn't see a reason to make it visiable in scaladoc?

File: scala-package/core/src/main/scala/ml/dmlc/mxnet/module/DataParallelExecutorGroup.scala [hunk 5]
```diff
@@ -0,0 +1,695 @@
...
+
+import ml.dmlc.mxnet.DType.DType
+import ml.dmlc.mxnet._
+import ml.dmlc.mxnet.module.DataParallelExecutorGroup.Builder
+import org.slf4j.{Logger, LoggerFactory}
+
+import scala.collection.mutable
+import scala.collection.mutable.ArrayBuffer
+
+object DataParallelExecutorGroup {
```
Comment (CodingCat) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/module/DataParallelExecutorGroup.scala]: private?

File: scala-package/core/src/main/scala/ml/dmlc/mxnet/module/DataParallelExecutorGroup.scala [hunk 6]
```diff
@@ -0,0 +1,695 @@
...
+      } else {
+        // negative axis means the there is no batch_size axis, and all the
+        // results should be the same on each device. We simply take the first one,
+        // without checking they are actually the same
+        tensors(0)
+      }
+    }
+  }
+
+  object Builder {
```
Comment (CodingCat) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/module/DataParallelExecutorGroup.scala]: private?

File: scala-package/core/src/main/scala/ml/dmlc/mxnet/module/DataParallelExecutorGroup.scala [hunk 7]
```diff
@@ -0,0 +1,695 @@
...
+          (k, if (fixedParamNames.contains(k)) "null" else gradReq)
+        } else if (dataNames.contains(k)) {
+          (k, if (inputsNeedGrad) gradReq else "null")
+        } else {
+          (k, "null")
+        }
+      }).toMap
+    }
+  }
+  class Builder(private val symbol: Symbol,
```
Comment (CodingCat) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/module/DataParallelExecutorGroup.scala]: private and a blank line above?

File: scala-package/core/src/main/scala/ml/dmlc/mxnet/IO.scala [hunk 8]
```diff
@@ -227,4 +228,27 @@ abstract class DataPack() extends Iterable[DataBatch] {
...
 }
 
+// Named data desc description contains name, shape, type and other extended attributes.
+case class DataDesc(name: String, shape: Shape,
+                    dtype: DType = Base.MX_REAL_TYPE, layout: String = "NCHW") {
+  override def toString(): String = {
+    s"DataDesc[$name,$shape,$dtype,$layout]"
+  }
+}
+object DataDesc {
```
Comment (CodingCat) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/IO.scala]: blank line above?

File: scala-package/core/src/main/scala/ml/dmlc/mxnet/module/BaseModule.scala [hunk 9]
```diff
@@ -0,0 +1,552 @@
...
+ *      this property is simply the *current* symbol being used. For other modules,
+ *      this value might not be well defined.
+ *
+ *  When those intermediate-level API are implemented properly, the following
+ *  high-level API will be automatically available for a module:
+ *
+ *  - `fit`: train the module parameters on a data set
+ *  - `predict`: run prediction on a data set and collect outputs
+ *  - `score`: run prediction on a data set and evaluate performance
+ * @author Yizhi Liu
```
Comment (CodingCat) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/module/BaseModule.scala]: BTW, author tags are not allowed in ASF projects

File: scala-package/core/src/main/scala/ml/dmlc/mxnet/module/BaseModule.scala [hunk 10]
```diff
@@ -0,0 +1,552 @@
...
+
+  private[mxnet] var binded: Boolean = false
+  private[mxnet] var forTraining: Boolean = false
+  private[mxnet] var inputsNeedGrad: Boolean = false
+  private[mxnet] var paramsInitialized: Boolean = false
+  private[mxnet] var optimizerInitialized: Boolean = false
+  private[mxnet] var symbol: Symbol = null
+  private[mxnet] var execGroup: DataParallelExecutorGroup = null
+  private[mxnet] var argParams: Map[String, NDArray] = null
+  private[mxnet] var auxParams: Map[String, NDArray] = null
```
Comment (CodingCat) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/module/BaseModule.scala]: a not-short list of package-private and mutable variables are fishy....any difficulty to make them immutable and minimize the scope?

File: scala-package/core/src/main/scala/ml/dmlc/mxnet/module/BaseModule.scala [hunk 11]
```diff
@@ -0,0 +1,552 @@
...
+          evalBatchEndCallback: Option[BatchEndCallback] = None,
+          initializer: Initializer = new Uniform(0.01f),
+          argParams: Map[String, NDArray] = null,
+          auxParams: Map[String, NDArray] = null,
+          allowMissing: Boolean = false,
+          forceRebind: Boolean = false,
+          forceInit: Boolean = false,
+          beginEpoch: Int = 0, numEpoch: Int = 1,
+          validationMetric: Option[EvalMetric] = None,
+          monitor: Option[Monitor] = None): Unit = {
```
Comment (CodingCat) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/module/BaseModule.scala]: ....fishy to have such long parameter list....define a parameter container class?

File: scala-package/core/src/main/scala/ml/dmlc/mxnet/module/DataParallelExecutorGroup.scala [hunk 12]
```diff
@@ -0,0 +1,695 @@
...
+import scala.collection.mutable
+import scala.collection.mutable.ArrayBuffer
+
+object DataParallelExecutorGroup {
+  private val logger: Logger = LoggerFactory.getLogger(classOf[DataParallelExecutorGroup])
+  // Load a list of arrays into a list of arrays specified by slices
+  private def loadGeneralMulti(data: Seq[NDArray],
+                               targets: Seq[Array[((Int, Int), NDArray)]],
+                               majorAxis: Seq[Int]): Unit = {
+    for (((dSrc, dTargets), axis) <- data zip targets zip majorAxis) {
```
Comment (CodingCat) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/module/DataParallelExecutorGroup.scala]: same here, we may want to consistent usage of operator style and method style?

Comment (CodingCat) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/module/DataParallelExecutorGroup.scala]: to make it easy to read, shall we have a meaningful variable for `data zip targets zip majorAxis`?

File: scala-package/core/src/main/scala/ml/dmlc/mxnet/module/BaseModule.scala [hunk 13]
```diff
@@ -0,0 +1,552 @@
...
+ *  When those intermediate-level API are implemented properly, the following
+ *  high-level API will be automatically available for a module:
+ *
+ *  - `fit`: train the module parameters on a data set
+ *  - `predict`: run prediction on a data set and collect outputs
+ *  - `score`: run prediction on a data set and evaluate performance
+ * @author Yizhi Liu
+ */
+abstract class BaseModule {
+  private val logger = LoggerFactory.getLogger(classOf[BaseModule])
```
Comment (Ldpe2G) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/module/BaseModule.scala]: What about make the logger to be null and only initialize in the subclass. Bacause from the training log I see BaseModule instead of Module.

Comment (CodingCat) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/module/BaseModule.scala]: I would propose to use something like Logging trait in Spark (https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/internal/Logging.scala), much easier to control

Comment (CodingCat) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/module/BaseModule.scala]: but that might be in another PR?

File: scala-package/examples/src/main/scala/ml/dmlc/mxnet/examples/module/MnistMlp.scala (hunk 3 above)
Comment (javelinjs) [inline scala-package/examples/src/main/scala/ml/dmlc/mxnet/examples/module/MnistMlp.scala]: Any suggestion? How does, e.g. Spark, deal with the convention?

Comment (CodingCat) [inline scala-package/examples/src/main/scala/ml/dmlc/mxnet/examples/module/MnistMlp.scala]: most of places in Spark uses method style....but there are also some places in Spark SQL to have inconsistent usage...so....no solution there...

File: scala-package/core/src/main/scala/ml/dmlc/mxnet/NDArray.scala (hunk 1 above)
Comment (javelinjs) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/NDArray.scala]: It is public. Do you mean we should make it private?

File: scala-package/examples/src/main/scala/ml/dmlc/mxnet/examples/module/MnistMlp.scala (hunk 2 above)
Comment (javelinjs) [inline scala-package/examples/src/main/scala/ml/dmlc/mxnet/examples/module/MnistMlp.scala]: removed

File: scala-package/core/src/main/scala/ml/dmlc/mxnet/module/DataParallelExecutorGroup.scala (hunk 12 above)
Comment (javelinjs) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/module/DataParallelExecutorGroup.scala]: Hmm... not quite understand 'have meaningful variable for `data zip targets zip majorAxis`'

File: scala-package/core/src/main/scala/ml/dmlc/mxnet/module/BaseModule.scala (hunk 9 above)
Comment (javelinjs) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/module/BaseModule.scala]: removed

File: scala-package/core/src/main/scala/ml/dmlc/mxnet/module/DataParallelExecutorGroup.scala (hunk 7 above)
Comment (javelinjs) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/module/DataParallelExecutorGroup.scala]: I'll make it `private[module]` since it is used in `Module.class`

File: scala-package/core/src/main/scala/ml/dmlc/mxnet/module/BaseModule.scala (hunk 10 above)
Comment (javelinjs) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/module/BaseModule.scala]: some of them are initialized when binding, and the bind state can be reset.

File: scala-package/core/src/main/scala/ml/dmlc/mxnet/NDArray.scala (hunk 1 above)
Comment (Ldpe2G) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/NDArray.scala]: It's strange, if it is public why I can't call this function. 
And is it the following code means making these functions private?https://github.com/dmlc/mxnet/blob/master/scala-package/macros/src/main/scala/ml/dmlc/mxnet/NDArrayMacro.scala#L59

Comment (javelinjs) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/NDArray.scala]: Sorry I forgot this. I thought function starts with "_" means that it should be private. @piiswrong

File: scala-package/core/src/main/scala/ml/dmlc/mxnet/module/DataParallelExecutorGroup.scala (hunk 12 above)
Comment (CodingCat) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/module/DataParallelExecutorGroup.scala]: I mean give a variable name for `data zip targets zip majorAxis`

Comment (javelinjs) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/module/DataParallelExecutorGroup.scala]: But we already have `((dSrc, dTargets), axis)` here.

Comment (CodingCat) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/module/DataParallelExecutorGroup.scala]: ah, sorry, I missed that part. sure

File: scala-package/core/src/main/scala/ml/dmlc/mxnet/module/BaseModule.scala [hunk 14]
```diff
@@ -0,0 +1,638 @@
...
+  def forwardBackward(dataBatch: DataBatch): Unit = {
+    forward(dataBatch, isTrain = Option(true))
+    backward()
+  }
+
+  /**
+   * Run prediction on `eval_data` and evaluate the performance according to `eval_metric`.
+   * @param evalData : DataIter
+   * @param evalMetric : EvalMetric
+   * @param numBatch Number of batches to run. Default is `None`,
```
Comment (Ldpe2G) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/module/BaseModule.scala]: Default is Integer.MAX_VALUE ?

File: scala-package/core/src/main/scala/ml/dmlc/mxnet/module/BaseModule.scala [hunk 15]
```diff
@@ -0,0 +1,638 @@
...
+ *
+ *  - `data_names`: list of string indicating the names of required data.
+ *  - `output_names`: list of string indicating the names of required outputs.
+ *
+ *  And also the following richer information after binded:
+ *
+ *  - state information
+ *    - `binded`: `bool`, indicating whether the memory buffers needed for computation
+ *      has been allocated.
+ *    - `for_training`: whether the module is binded for training (if binded).
```
Comment (Ldpe2G) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/module/BaseModule.scala]: Should be forTraining , same for the following.

File: scala-package/core/src/main/scala/ml/dmlc/mxnet/module/BaseModule.scala (hunk 14 above)
Comment (javelinjs) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/module/BaseModule.scala]: fixed

File: scala-package/core/src/main/scala/ml/dmlc/mxnet/module/BaseModule.scala (hunk 15 above)
Comment (javelinjs) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/module/BaseModule.scala]: fixed

**Comment by javelinjs** [comment]
If everything's right, I'll do the merge. @CodingCat @Ldpe2G
